#### Clarusway Python

* [Instructor Landing Page](landing_page.ipynb)
* <a href="https://colab.research.google.com/github/4dsolutions/clarusway_data_analysis/blob/main/basic_python/19.Python_Session19.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open and Execute in Google Colaboratory"></a>
* [![nbviewer](https://raw.githubusercontent.com/jupyter/design/master/logos/Badges/nbviewer_badge.svg)](https://nbviewer.org/github/4dsolutions/clarusway_data_analysis/blob/main/basic_python/19.Python_Session19.ipynb)

<a id="toc"></a>

## <p style="background-color:#0D8D99; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">Python Session 19</p>

#### <div class="alert alert-block alert-info"><h1><p style="text-align: center; color:purple"> Packages & Modules<br><br>Comparing "list comprehension" vs. "for loop" performance<br>with timeit module.<br><br>Algorithm Exercise-The factorial of a number using Recursive Function</p> 

**Python official document - Python Packaging User Guide: https://packaging.python.org/en/latest/**

**Python official web site - Find, install and publish Python packages with the Python Package Index: https://pypi.org/**

**Python official document - Python Module Index: https://docs.python.org/3/py-modindex.html**

**Python official document - Python Standart Library :https://docs.python.org/3/library/**

**Pip documentation - Pip commands: https://pip.pypa.io/en/stable/cli/index.html**

<a id="toc"></a>

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">Packages & Modules</p>

**In Python, a script is a program that typically performs a specific function and is contained in a single file. Scripts are useful when code will not be reused or used in another project.**

**A module, on the other hand, is a reusable block of code that can be used in different scripts. Modules can contain elements such as functions, classes, and constants. Modules enable code reuse and are particularly useful in large projects. Modules can be used in another script or module using the import keyword.**

In [1]:
def square(a) :
    return a * a

print(square(3))

9


**Here, we are seeing the usage of a module inside a script by importing it:**

In [2]:
import math

def square(a) :
    return a * a

print(square(3))

math.sqrt(16)

9


4.0

In [3]:
pwd  # see the absolute path of your working directory

'/Users/kirbyurner/Documents/clarusway_data_analysis/basic_python'

In [4]:
# Let's import the math module.

import math

In [5]:
# math.  # after the dot, press TAB and see the module's functions

In [6]:
# Now that I've imported it, I can see what's inside.

print(dir(math))

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'cbrt', 'ceil', 'comb', 'copysign', 'cos', 'cosh', 'degrees', 'dist', 'e', 'erf', 'erfc', 'exp', 'exp2', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'isqrt', 'lcm', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'nextafter', 'perm', 'pi', 'pow', 'prod', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc', 'ulp']


In Python, there are special objects that start and end with double underscores in modules. These objects are used to exhibit special behaviors and are referred to as "special dunder (double underscore) methods" in the Python language.

These objects can be used for many different purposes. For example, the init method is automatically called when a class instance is created and contains the code that sets up the instance's initial state. Similarly, the len method contains the code that calculates the length of an object, and the str method contains the code that creates a string representation of an object.

These special objects are used by many different data types and classes in Python. These include particularly classes, attributes, methods, and modules.

Exactly right! The init method is the first method called automatically when a class instance is created and is used to initialize instance attributes. Another use of special methods is to determine how objects perform arithmetic operations. For example, the add method can be used for the + operator, the sub method for the - operator, and so on.

Additionally, special methods can be used to provide information about an object's contents and behavior. The len method is used to calculate the length of an object, while the str method is used to create a string representation of an object. Similarly, the repr method can be used to create a string representation of an object that is a descriptive character string.

Special methods are an essential part of the Python programming language and are used frequently when writing Python code. When used correctly, they can make code more readable and understandable. However, when used unnecessarily, they can make code more complex.

In [7]:
print(math.__spec__)

ModuleSpec(name='math', loader=<_frozen_importlib_external.ExtensionFileLoader object at 0x10e38dc10>, origin='/Users/kirbyurner/opt/anaconda3/envs/py311/lib/python3.11/lib-dynload/math.cpython-311-darwin.so')


In [8]:
from math import pi, factorial, log10

In [9]:
pi

3.141592653589793

In [10]:
factorial(5)

120

In [11]:
def my_factorial(n) :
    sonuç = 1

    for i in range(1, n+1) :
        sonuç *= i        

    if n == 0 or n == 1 :
        return 1    
    else : return sonuç

In [12]:
log10(1000)

3.0

**Task :**

**Let’s import punctuation and digits function from string module**

**Print the all punctuation marks and digits chars using these functions.**

In [13]:
import string as stg

In [14]:
print(stg.punctuation)

# This is a variable containing all punctuation marks

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [15]:
stg.digits

# this is a variable containing all digits.

'0123456789'

In [16]:
# Let's see what else is in the string module?

dir(stg)

['Formatter',
 'Template',
 '_ChainMap',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_re',
 '_sentinel_dict',
 '_string',
 'ascii_letters',
 'ascii_lowercase',
 'ascii_uppercase',
 'capwords',
 'digits',
 'hexdigits',
 'octdigits',
 'printable',
 'punctuation',
 'whitespace']

## Task :

**Let’s import today function from date object and now function from datetime object all from datetime module,**

**Print the current date (yyyy-mm-dd) and time using these functions.**

In [17]:
import datetime

In [18]:
print(datetime.date.today())

2024-11-21


In [19]:
print(datetime.datetime.now())

2024-11-21 14:37:14.916972


## Task : 

**Using datetime module, write a program to calculate the following:**

**According to the general acceptance, Ali was born on 22th April 571 AD, and died on 8th June 632 AD.** 

**How many days have he lived in his life?**

In [20]:
from datetime import date

In [24]:
date(2024, 11, 21) # shift tab tab --> year, month, day

datetime.date(2024, 11, 21)

In [25]:
birth = date(1809, 2, 12)

# birth = date(571, 4, 22)

In [ ]:
type(date(1809, 2, 12))

# type(date(571, 4, 22))

In [ ]:
death = date(1865, 4, 15)

# death = date(632, 6, 8)

In [ ]:
death - birth

In [ ]:
print(death - birth)

In [ ]:
date.toordinal(death)

In [ ]:
date.toordinal(birth)

In [ ]:
life_day = date.toordinal(death) - date.toordinal(birth)
life_day

In [ ]:
from datetime import date

birth = date(1809, 2, 12)
death = date(1865, 4, 15)

life_day = date.toordinal(death) - date.toordinal(birth)
life_day

## "random" is a module that contains functions that allow us to select randomly from various data types.

## Task :

**Let’s import choice function from random module,**

**Print one of the element of the following list randomly.**

In [ ]:
city = ["stockholm", "istanbul", "seul", "cape town"]

In [ ]:
from random import choice

In [ ]:
choice(city)

## Package Managers:  conda and pip

In [29]:
# Let's look at the version of our conda utility

! conda --version

conda 24.7.1


In [26]:
# let's import numpy module.

import numpy

In [ ]:
# !pip install numpy  DON'T DO THIS 

In [30]:
#  Let's see what libraries my jupyter can run.

# ! conda list # might take awhile

In [33]:
# we can display the information of a module on the screen.

! conda info


     active environment : base
    active env location : /Users/kirbyurner/opt/anaconda3/envs/py311
            shell level : 1
       user config file : /Users/kirbyurner/.condarc
 populated config files : /Users/kirbyurner/.condarc
          conda version : 24.7.1
    conda-build version : not installed
         python version : 3.11.3.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=broadwell
                          __conda=24.7.1=0
                          __osx=12.7.6=0
                          __unix=0=0
       base environment : /Users/kirbyurner/opt/anaconda3/envs/py311  (writable)
      conda av data dir : /Users/kirbyurner/opt/anaconda3/envs/py311/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/osx-64
                          https:

In [34]:
! conda list numpy

# packages in environment at /Users/kirbyurner/opt/anaconda3/envs/py311:
#
# Name                    Version                   Build  Channel
numpy                     1.26.4          py311hc43a94b_0    conda-forge
numpydoc                  1.8.0              pyhd8ed1ab_0    conda-forge


<a id="toc"></a>

### <p style="background-color:#0D8D99; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">Comparing "list comprehension" vs. "for loop" performance<br> with timeit module.</p>

When we research, we come across information that list comprehension provides us with a speed advantage. But have you ever thought about how we can test this?

List comprehension expands a list for us with a loop and returns a list in the end. So, if I create a list using both "for loop" and list comprehension and measure the time it takes for both of them to complete, I will have tested how much faster list comprehension is.

Are you curious about how much difference there will be at the millisecond level?
Let's do this together:

In [ ]:
from timeit import timeit 

In [ ]:
timeit() # shift tab tab

We can put a function inside the timeit function and run it one million times. timeit keeps track of its duration and produces a timer object. We obtain the elapsed time in milliseconds.

**First, let's define a function that uses a classic for loop inside. This function should generate a number series and add them up in a list:**

In [ ]:
def for_loop():
    result = [] 
    
    for i in range(1000000) : # Let's loop it 1 million times.
        result.append(i)
    return result  # finally return my list.

**Then, let's define another function that does the same job using a list comprehension.**

In [ ]:
def list_comp() :
    return [i for i in range(1000000)]

**Now let's compare these two functions with the timeit function:**

In [ ]:
size = 100

time_for = timeit(for_loop, number = size)
# This is the time for the for loop to complete the operation

time_list_comp = timeit(list_comp, number = size)
# This is the time for the list comprehension to complete the operation

**Let's print this comparison to the screen in an understandable way:**

In [ ]:
print(f"List Comp. is {round(time_for/time_list_comp, 2)} times faster than for loops.")

# <a id="toc"></a>

### <p style="background-color:#0D8D99; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">An Algorithm Exercise</p>

<a id="toc"></a>

### <p style="background-color:#9d4f8c; font-family:newtimeroman; color:#FFF9ED; font-size:175%; text-align:center; border-radius:10px 10px;">Recursivity</p>

In Python, recursive functions are used to solve a problem by calling themselves. These types of functions are commonly used to work with repeating structures such as tree structures or recursive mathematical operations.

Recursive functions stop calling themselves and return when a condition is reached, which is sometimes referred to as the **"base case"**. The main function of a recursive function is to call itself until the base case is reached, and then combine the results.

Recursive functions often help you create simpler and more understandable code. However, they can cause infinite loops when misused or conditioned incorrectly.

**Recursive functions are typically defined like this:**

In [ ]:
def recursive_function(parameters):
    if base_case_condition:
        return base_case_value
    else:
        recursive_function(modified_parameters)

Here, base_case_condition is the condition that determines the base case, and base_case_value is the value that will be returned at the base case.

modified_parameters are the parameters that are modified when the function is called again.

**For example, when we want to write a code to find the factorial of a number, we can easily solve it using a recursive function**

<hr>

Let's first write a program using the traditional way to find the factorial of a number. Then, let's solve the same problem using recursion function and compare the difference between them.
<hr>

<hr>

The factorial of a positive integer is the product of all positive integers starting from 1 up to and including that number.

For example, the factorial of 5 is 1 * 2 * 3 * 4 * 5 = 120.
<hr>

<hr>

When defining our algorithm, we need to find a connection or pattern between the result and the numbers involved. 

Multiplying all numbers between 1 and the given number means multiplying that number by one less each time. Yes, I noticed this relationship. 

So, I need to use a number sequence here. If I'm going to use a number sequence, I immediately think of the range() function.

Yes, I need the range function because if I'm going to find the factorial of 5, I will need all numbers from 1 to 5 inclusive. If I'm going to find the factorial of 10, I will need all numbers from 1 to 10 inclusive. I will multiply them with each other.
<hr>

If I'm going to use each element of this number array, then a for loop is winking at me here. 

We are talking about element-wise iteration because we will take one element at each iteration and perform a mathematical operation. 

In each iteration of the loop, we will multiply the next element with the previous one and update the result by multiplying it with the next element in the next iteration.

In [ ]:
def my_factorial(n) :  
    result = 1 
    for i in range(1, n+1) : 
        result *= i
    return result

In [ ]:
my_factorial(5)

**The factorial of 0 and 1 is also 1.**

If we think that there are exceptions in our code, we should think of **"conditions"**. 

**I should use an if statement so that the result returns as I have specified under certain conditions**

In [ ]:
def my_factorial(n) :  
    result = 1 
    for i in range(1, n+1) : 
        result *= i
    if n == 0 or n == 1 :  # If the value of n is 0 or 1, return the result as 1.
        return 1 
    else :  # If the value of n is not 0 or 1, return whatever the result is.
        return result

In [ ]:
my_factorial(1)

### Now let's achieve the same solution using a recursive function:

**Since a recursive function calls itself, there is a risk of entering an infinite loop; therefore, YOU MUST USE A CONDITION.**

We write recursive functions in two conditions:

The first condition is the **"base case"** condition in which we determine under which condition the recursion will end.

The second condition is the **"recursive case"** condition that contains the general solution of the function that we call itself

In [ ]:
def factorial(n) :  
    if n == 1 or n == 0 : # This is the 'base case', 
                # and if the value of n is 1 or 0, the code block of this if statement will run, 
                # and the function will terminate.
        return 1  
    else :  # This is the condition that contains the general solution, 
        # and the case that contains the general solution should be here.
        return n * factorial(n-1)

**Let's try to understand the recursive operation we did with the following diagram:**

Let's assume that we are calculating the factorial of 3 instead of 5 to simplify the explanation:

The function will be initially called with 3 as its argument. When the Python interpreter reads the code block of **else**, it will execute the code **"3 multiplied by factorial(3-1)"**. 

Therefore, it will not return a result at this point because the factorial() function will be called with the argument "n-1", which is 2.

Unlike in the for loop, the interpreter does not go back to the beginning. 

When the function is called with 2 as its argument, this time the code block of **else** will run with the code **"3 multiplied by 2 multiplied by factorial(2-1)"**. 

So, this time the factorial() function will be called with 1 as its argument. **When the function is called with 1, the 'base case' condition will occur, and the if condition will return True, and the recursive function will terminate by returning 1.**

**At the end of all these operations, what will be returned is 3 times 2 times 1. Therefore, it will return the factorial of 3.**

## Note! 

This function calculates factorial using **base case** checks and **recursive** calls. 

**However, because recursive functions tend to run slower and use more memory, it is not the preferred method for simple operations like factorial.**

### Let's compare performance of recursive and normal functions with factorial calculation :

In [ ]:
import time

# Recursive fonksiyon ile faktöriyel hesabı
def factorial_recursive(n):
    if n == 0 or n == 1:
        return 1
    else:
        return n * factorial_recursive(n-1)

# Iterative fonksiyon ile faktöriyel hesabı
def factorial_iterative(n):
    result = 1
    for i in range(1, n+1):
        result *= i
    return result

# Zaman ölçümü
start_time = time.time()
factorial_recursive(500)
end_time_recursive = time.time()
time1 = round(((end_time_recursive - start_time)*1000), 4)

print("Recursive function execution time: ", time1)

start_time = time.time()
factorial_iterative(500)
end_time_iterative = time.time()
time2 = round(((end_time_iterative - start_time)*1000), 4)

print("Iterative function execution time: ", time2)

print("Recursive function is {} times slower than iterative function".format(time1/time2))

 # Some examples of solutions with recursive functions :
 
 ## Please review these codes in your post-class studies!

### Reverse a string

In [ ]:
def reverse(string):
    if len(string) == 0:
        return string
    else:
        return reverse(string[1:]) + string[0]


reverseme = 'Desserts'
print(reverse(reverseme))

### Fibonacci Sequence

In [ ]:
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)


number = 11

print('Fibonacci sequence:')
for i in range(number):
    print(fibonacci(i), end =" ")

### "n"th fibbonacci number

In [ ]:
def fibonacci(nthNumber):
    if nthNumber == 1 or nthNumber == 2:
        # BASE CASE
        return 1
    else:
        # RECURSIVE CASE
        result = fibonacci(nthNumber - 1)
        result = result + fibonacci(nthNumber - 2)
        return result

print(fibonacci(10))

###  The sum of numbers from 1 to n like 1 + 2 + 3 + 4 +, etc.

In [ ]:
def sumnums(n):
    if n == 1:
        return 1
    return n + sumnums(n - 1)


print(sumnums(5))

### Count down and up

In [ ]:
def countDownAndUp(number):
    print(number)
    if number == 0:
        # BASE CASE
        print('Reached the base case.')
        return
    else:
        # RECURSIVE CASE
        countDownAndUp(number - 1)
        print(number)
        return

countDownAndUp(3)